src/pointerwrappers.jl

```diff
@@ -58,10 +58,12 @@ Base.propertynames(::PointerWrapper{T}) where T = fieldnames(T)
 function Base.getproperty(pw::PointerWrapper{T}, name::Symbol) where T
   i = findfirst(isequal(name), fieldnames(T))
   if isnothing(i)
-    error("type $(string(T)) has no field $name")
+    # For some `struct`s, `fieldnames` gives `data` and not the actual field names, but we can use `Base.getproperty` for pointers,
+    # see https://github.com/trixi-framework/P4est.jl/issues/72
+    return PointerWrapper(Base.getproperty(pointer(pw), name))
   end
 
-  PointerWrapper(fieldtype(T, i), pointer(pw) + fieldoffset(T, i))
+  return PointerWrapper(fieldtype(T, i), pointer(pw) + fieldoffset(T, i))
 end
```

Comment:
I wonder if we should try to use a `@generated` function, now that the logic is becoming more and more complicated here. Maybe something like this:
```julia
# Syntactic sugar: allows one to use `pw.fieldname` to get a PointerWrapper-wrapped pointer to `fieldname`
@generated function Base.getproperty(pw::PointerWrapper{T}, name::Symbol) where T
  i = findfirst(isequal(name), fieldnames(T))
  if isnothing(i)
    # For some `struct`s, `fieldnames` gives `data` and not the actual field names, but we can use `Base.getproperty` for pointers,
    # see https://github.com/trixi-framework/P4est.jl/issues/72
    return quote
      $(Expr(:meta, :inline))
      PointerWrapper(Base.getproperty(pointer(pw), name))
    end
  else
    return quote
      $(Expr(:meta, :inline))
      PointerWrapper($(fieldtype(T, i)), pointer(pw) + $(fieldoffset(T, i)))
    end
  end
end
```
But don't assume I got the macro syntax correct 😅

Reply:
I think this needs some debugging. Note that you have only access to the types when using a `@generated` function, see https://docs.julialang.org/en/v1/manual/metaprogramming/#Generated-functions
Thus, I would expect that we can only see that `name` is a `Symbol` when calling such a `@generated` function - while we would like to use its value...

Reply:
I've seen this used elsewhere before, where they encoded this in `Val{...}`s and then were actually able to access the type in `@generated` functions. However, this seems overkill here, so let's go down that road only if and when necessary.